cvc5_pythonic_api/cvc5_pythonic.py

```diff
@@ -1727,6 +1735,544 @@ def Or(*args):
     return _nary_kind_builder(Kind.OR, *args)
 
 
+#########################################
+#
+# Strings and Sequences
+#
+#########################################
+
+class SeqSortRef(SortRef):
+    def cast(self,val):
+        """Try to cast `val` as a String.
+        """
+        if is_expr(val):
+            if debugging():
+                _assert(self.ctx == val.ctx, "Context mismatch")
+            val_s = val.sort()
+            if self.eq(val_s):
+                return val
+            if val_s.is_bool() and self.is_int():
+                return If(val, 1, 0)
+            if val_s.is_bool() and self.is_real():
+                return ToReal(If(val, 1, 0))
+            if val_s.is_int() and self.is_real():
+                return ToReal(val)
+            if debugging():
+                _assert(False, "SMT Integer/Real expression expected")
+        elif type(val) == str:
+            return StringVal(str(val))
+        else:
+            return None
+    
+    def is_string(self):
+        """Determine if sort is a string
+        >>> s = StringSort()
+        >>> s.is_string()
+        True
+        >>> s = SeqSort(IntSort())
+        >>> s.is_string()
+        False
+        """
+        return isinstance(self,StringSortRef)
+    
+class SeqRef(ExprRef):
+    """Sequence Expressions"""
+    def sort(self):
+        return _sort(self.ctx, self.ast)
+    
+    def isSequence(self):
+        return True
+    
+    def __add__(self, other):
+        return Concat(self, other)
+    
+    def __radd__(self, other):
+        return Concat(other, self)
+    
+    def __getitem__(self, i):
+        if _is_int(i):
+            i = IntVal(i, self.ctx)
+        return _to_expr_ref(self.ctx.solver.mkTerm(Kind.SEQ_NTH,self.ast,i.ast),self.ctx)
+
+    def at(self, i):
+        if _is_int(i):
+            i = IntVal(i, self.ctx)
+        return  SeqRef(self.ctx.solver.mkTerm(Kind.SEQ_AT, self.ast,i.ast),self.ctx)
+
+    def is_string(self):
+        return isinstance(self,StringRef)
+    
+
+
+
+class StringSortRef(SeqSortRef):
```

Comment:
The main reason it's like this is that I started with strings and then added sequences. Since strings are a kind of sequences it made sense to add another class for sequences and then string inherits from sequence.

Reply:
I'll add a couple more comments about this:

1. z3 seems to treat strings as a particular case of sequences, where the elements are `characters`. In contrast, the SMTLIB standard (and cvc5) do not have a `character` sort, but instead strings are identified as characters of length 1. Hence it makes sense to separate them. Still, as Laor mentioned, there is some sharing, which can be nicely implemented using inheritence.
2. In z3's SeqSortRef class, they have a flag `is_string`. I think it is nicer (at least from an object oriented perspective) to use inheritence instead.
3. Most importantly, if I understand correctly, all of this should be transparent to the user, as the difference is only in the `SeqSortRef` class, and not, e.g., in `SeqSort` or `StringSort`, that are also separated in z3.


If you agree with this design, would something along the lines of the above explanation (just shorter) be suitable in the documentation?

Reply:
Yep, this makes sense! I'd stash this in a comment somewhere. I don't think we need it in the external documentation.

Reply:
@yoni206 : just bumping this.

Reply:
@LaorS , @alex-ozdemir and I talked about this offline -- can you plz write a comment about this in the code, for future developers?